doc: gsg cleanups (#23977)
zephyrproject-rtos/zephyr: doc/getting_started/index.rst

Getting Started Guide
#####################

Follow this guide to get a quick start with Zephyr development where
you'll:
Follow this guide to:

- **Set up a command-line development environment** for Linux* (Ubuntu),
  macOS, or Windows, with required package manager, compiler, and
  build-system tools,
- **Get the sources**,
- **Build, flash, and run** a sample application on your target board.

> \*Instructions for other Linux distributions are discussed in the

Author: OK, I'll remove it from the first ubuntu tab instead, then. The same link appears twice in consecutive paragraphs, so I was trying to remove one of them. I agree keeping it here makes more sense.

Select and Update OS
********************

> Click the operating system you are using.

Reviewer: When looking at the rendered doc, I think there should be just a single line explaining why users should do those commands, so maybe add a little:
```suggestion
Click the operating system you are using and follow the guide to ensure it is up-to-date.
```

Author: I think this is implied by the title "Select and Update OS"

Reviewer: yeah, but when I first read it, I just read the content and my eyes didn't notice the title.
Not sure if i'm the only one :laughing:

Author: I guess adding more text isn't likely to make eyes glaze less, though. That's exactly why I'm trying to strip this down and remove any text I can -- people generally ignore text and just start copy pasting command lines into their terminals as fast as they can, in my experience.

Install dependencies
********************

Next, you'll install some host dependencies using your package manager.
You'll also install :ref:`west <west>`, Zephyr's multi-purpose tool.

.. tabs::

   .. group-tab:: Ubuntu

      #. Use ``apt`` to install dependencies:

         .. code-block:: bash

            sudo apt install --no-install-recommends cmake ninja-build gperf \
              python3 python3-setuptools python3-pip python3-wheel git \
              ccache wget xz-utils file

      #. Verify the version of cmake installed on your system using::

            cmake --version

         If it's not version 3.13.1 or higher, follow these steps to
         add the `Kitware third-party apt repository <https://apt.kitware.com/>`__
         to get an updated version of cmake.

         a) Add the Kitware signing key:

            .. code-block:: bash

               wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | sudo apt-key add -

         b) Add the Kitware apt repository for your OS release. For Ubuntu

>             18.04:

Reviewer: I think `Ubuntu 18.04 LTS` is the right term: https://releases.ubuntu.com/

         c) Then install the updated cmake with ``apt``:

            .. code-block:: bash

               sudo apt update
               sudo apt install cmake

      #. Install west, and make sure :file:`~/.local/bin` is on your
         :envvar:`PATH` :ref:`environment variable <env_vars>`:

         .. code-block:: bash

            pip3 install --user -U west
            echo 'export PATH=~/.local/bin:"$PATH"' >> ~/.bashrc
            source ~/.bashrc

   .. group-tab:: macOS

      #. Install `Homebrew <https://brew.sh/>`_:

         .. code-block:: bash

            /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

      #. Use ``brew`` to install dependencies:

         .. code-block:: bash

            brew install cmake ninja gperf python3 ccache

      #. Install west:

         .. code-block:: bash

            pip3 install west

   .. group-tab:: Windows

      .. note:: We don't recommend using the `Windows Subsystem for Linux (WSL)
         <https://msdn.microsoft.com/en-us/commandline/wsl/install_guide>`__
         for Zephyr development yet, because it's harder to flash Zephyr

>          applications to boards using WSL.

Reviewer: Not supported, or just harder ?
If it is just harder, should we make a page describing, how to do it ?

Author: A user reported it's possible if you manually rename some of the underlying flash utilities. I haven't figured it out or even tried. I think that's future work and the recommendation should stand; I'm just trying to avoid a potentially inaccurate claim that it "can't" be done since users are claiming it can.

Reviewer: Ok, thanks for the explanation.
Then howabout something like:

```
The Zephyr project does not support flashing of your application using the
built-in Windows Subsystem for Linux (WSL).

Therefore we don't recommend using the `Windows Subsystem for Linux (WSL).
```
that way we only say we are not supporting it, which is different from: It cannot be done.

Install dependencies
********************

Next, you'll install some host dependencies using your package manager.

.. tabs::

   .. group-tab:: Ubuntu

      #. Use ``apt`` to install dependencies:

         .. code-block:: bash

            sudo apt install --no-install-recommends cmake ninja-build gperf \
              python3 python3-setuptools python3-pip python3-wheel git \

>               ccache wget xz-utils file

Reviewer 2: native posix won't work if you don't get the multilibs. Are you sure we want to remove those?

Author: How about we move those to the toolchain page in that case?

Reviewer: I think the getting started guide should get you all needed tools without looking elsewhere.

And being able to download Zephyr and just test it by building native posix, is one such thing that should be possible out-of-the box.

Just as it makes life easier for Linux users, that Zephyr provides a SDK with toolchains, so you don't need to go to 3rd party toolchain page (which you need to do on windows)

Reviewer 3: > How about we move those to the toolchain page in that case?

how is this related to toolchains? Agree with @tejlmand

Author: Those are the toolchains that are needed to build a particular target (native_posix)... what am I missing?

Reviewer: and so is: `arm`, `arc`, `xtensa`, and all the other toolchains.
Why should they work, and not `native_posix`, when we don't know which toolchain the users will be using with Zephyr ?

Reviewer 2: Agreed, I suggest leaving the multilib requirements as they were. native_posix is very important to many newcomers to Zephyr.

Author: I've restored the Ubuntu packages and moved the Python script changes to their own PR, https://github.com/zephyrproject-rtos/zephyr/pull/24248

Install dependencies
********************

> Next, you'll install some host dependencies using your package manager.

Reviewer: Maybe emphasize the importance ?
```suggestion
Next, you'll install required host dependencies using your package manager.
```

Author: Well, they are not all required for all use cases, as discussed earlier in the review. Strictly speaking several of these are optional dependencies only needed for some targets or some use cases.

Since I was requested to leave all the optional dependencies I discovered in thisdocument, I have done so, but they're not required ;).

Author: For reference and the record:

- ccache: this is an optimization
- wget: only needed to pull down the Zephyr SDK, which is optional
- make: optional
- gcc, gcc-multilib, g++-multilib: only needed for some host targets, per review
- dfu-util: only needed for some boards as a runner dependency
- libsdl2-dev: I honestly don't know
- python3-tk: only needed for the guiconfig target, as far as I can tell
- file: only needed by the SDK installer
- xz-utils: needed to unpack the SDK, not required on all toolchains
- device-tree-compiler: optional, just used for additional warnings

You can see that not much is actually required.

Reviewer: Thanks for the list.
Could we add this information somewhere and link to it ?

Personally I like to be able to know why I need to install a certain tool, if it is not strictly required. 
(Note: this can be an independent improvement to the cleanup)

Author: I'd rather not maintain such a thing and it's likely to go stale. People tend to build additional users on existing dependencies.

Reviewer: point taken.

            18.04 LTS:

            .. code-block:: bash

               sudo apt-add-repository 'deb https://apt.kitware.com/ubuntu/ bionic main'

         c) Then install the updated cmake with ``apt``:

            .. code-block:: bash

               sudo apt update
               sudo apt install cmake

   .. group-tab:: macOS

      #. Install `Homebrew <https://brew.sh/>`_:

         .. code-block:: bash

            /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

      #. Use ``brew`` to install dependencies:

         .. code-block:: bash

            brew install cmake ninja gperf python3 ccache

   .. group-tab:: Windows

      .. note::

         Due to issues finding executables, the Zephyr Project doesn't
         currently support application flashing using the `Windows Subsystem
         for Linux (WSL)
         <https://msdn.microsoft.com/en-us/commandline/wsl/install_guide>`_
         (WSL).

         Therefore, we don't recommend using WSL when getting started.

      These instructions must be run in a ``cmd.exe`` command prompt. The
      required commands differ on PowerShell.

>       They rely on `Chocolatey`_. If Chocolatey isn't an option, you can

Reviewer 2: ```suggestion
      The instructions rely on `Chocolatey`_. If Chocolatey isn't an option, you can
```

            cmake --version

         If it's not version 3.13.1 or higher, follow these steps to
         add the `Kitware third-party apt repository <https://apt.kitware.com/>`__
         to get an updated version of cmake.

         a) Add the Kitware signing key:

            .. code-block:: bash

               wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | sudo apt-key add -

         b) Add the Kitware apt repository for your OS release. For Ubuntu
            18.04 LTS:

            .. code-block:: bash

               sudo apt-add-repository 'deb https://apt.kitware.com/ubuntu/ bionic main'

         c) Then install the updated cmake with ``apt``:

            .. code-block:: bash

               sudo apt update
               sudo apt install cmake

   .. group-tab:: macOS

      #. Install `Homebrew <https://brew.sh/>`_:

         .. code-block:: bash

            /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

      #. Use ``brew`` to install dependencies:

         .. code-block:: bash

>             brew install cmake ninja gperf python3 ccache

Author: That was a leftover from removing optional dependencies. Fixed, thanks.

      #. Install `udev <https://en.wikipedia.org/wiki/Udev>`_ rules, which
         allow you to flash most Zephyr boards as a regular user:

         .. code-block:: bash

            sudo cp ${ZEPHYR_SDK_INSTALL_DIR}/sysroots/x86_64-pokysdk-linux/usr/share/openocd/contrib/60-openocd.rules /etc/udev/rules.d
            sudo udevadm control --reload

   .. group-tab:: macOS

      Follow the instructions in :ref:`gs_toolchain`. Note that the Zephyr SDK
      is not available on macOS.

      Do not forget to set the required :ref:`environment variables <env_vars>`
      (:envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and toolchain specific ones).

   .. group-tab:: Windows

      Follow the instructions in :ref:`gs_toolchain`. Note that the Zephyr SDK
      is not available on Windows.

      Do not forget to set the required :ref:`environment variables <env_vars>`
      (:envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and toolchain specific ones).

.. _getting_started_run_sample:

.. rst-class:: numbered-step

Build the Blinky Sample
***********************

.. note::

   Blinky is compatible with most, but not all, :ref:`boards`. If your board

>    does not meet Blinky's :ref:`blinky-sample-requirements`, :ref:`hello_world`

Reviewer 2: ```suggestion
   does not meet Blinky's :ref:`blinky-sample-requirements`, then :ref:`hello_world`
```

Reviewer 2: This is so that the 2 hyperlinks do not "visually merge"

         west build -p auto -b <your-board-name> samples/basic/blinky

   .. group-tab:: Windows

      .. code-block:: bat

         cd %HOMEPATH%\zephyrproject\zephyr
         west build -p auto -b <your-board-name> samples\basic\blinky

The ``-p auto`` option automatically cleans byproducts from a previous build
if necessary, which is useful if you try building another sample.

.. rst-class:: numbered-step

Flash the Sample
****************

Connect your board, usually via USB, and turn it on if there's a power switch.
If in doubt about what to do, check your board's page in :ref:`boards`.

Then flash the sample using :ref:`west flash <west-flashing>`:

.. code-block:: console

   west flash

You may need to install additional :ref:`host tools <debug-host-tools>`
required by your board. The ``west flash`` command will print an error if any
required dependencies are missing.

If you're using blinky, the LED will start to blink as shown in this figure:

.. figure:: img/ReelBoard-Blinky.gif
   :width: 400px
   :name: reelboard-blinky

   Phytec :ref:`reel_board <reel_board>` running blinky

Next Steps
**********

Here are some next steps for exploring Zephyr:

> * Try other :ref:`samples-and-demos`.

Reviewer 2: I suggest also linking to the debugging section here, which is likely one of the things users want to try first.

            choco feature enable -n allowGlobalConfirmation

      #. Use ``choco`` to install dependencies:

         .. code-block:: console

            choco install cmake --installargs 'ADD_CMAKE_TO_PATH=System'
            choco install ninja gperf python git

      #. Open a new ``cmd.exe`` window **as a regular user** to continue.

.. _Chocolatey: https://chocolatey.org/
.. _Install chocolatey: https://chocolatey.org/install

.. _get_the_code:
.. _clone-zephyr:
.. _install_py_requirements:
.. _gs_python_deps:

.. rst-class:: numbered-step

> Install Zephyr

Author: > we do not want to make it sound as if Zephyr was an installable application, because it is not.

addressed

      #. Open a new ``cmd.exe`` window **as a regular user** to continue.

.. _Chocolatey: https://chocolatey.org/
.. _Install chocolatey: https://chocolatey.org/install

.. _get_the_code:
.. _clone-zephyr:
.. _install_py_requirements:
.. _gs_python_deps:

.. rst-class:: numbered-step

Install Zephyr
**************

Next, install Zephyr and its :ref:`modules <modules>` into a new :ref:`west
<west>` workspace named :file:`zephyrproject`. You'll also install Zephyr's
additional Python dependencies.

.. tabs::

   .. group-tab:: Ubuntu

      #. Install west, and make sure :file:`~/.local/bin` is on your
         :envvar:`PATH` :ref:`environment variable <env_vars>`:

         .. code-block:: bash

            pip3 install --user -U west
            echo 'export PATH=~/.local/bin:"$PATH"' >> ~/.bashrc
            source ~/.bashrc

      #. Get the Zephyr source code:

         .. code-block:: bash

            west init ~/zephyrproject
            cd ~/zephyrproject
            west update

>       #. Export a :ref:`Zephyr CMake package <cmake_pkg>`. This allows CMake to
>          load boilerplate code required for building Zephyr applications.

Reviewer: Changing the changes in later commit in same PR.
So unfriendly of you :wink: 

```suggestion
      #. Export a :ref:`Zephyr CMake package <cmake_pkg>`. This allows CMake to
         automatically load boilerplate code required for building Zephyr applications.
```

Author: Sorry, I will try to be more friendly in the future. Comments now addressed.

            choco feature enable -n allowGlobalConfirmation

      #. Install CMake:
      #. Use ``choco`` to install dependencies:

         .. code-block:: console

            choco install cmake --installargs 'ADD_CMAKE_TO_PATH=System'
            choco install ninja gperf python git

      #. Install the rest of the tools:

         .. code-block:: console

>             choco install git python ninja dtc-msys2 gperf

Reviewer 4: @mbolivar-nordic I'm sure this is documented somewhere (as otherwise we'd have new users complaining all day long), but I wonder how windows users are supposed to get dtc if not documented here.

Author: > @mbolivar-nordic I'm sure this is documented somewhere (as otherwise we'd have new users complaining all day long), but I wonder how windows users are supposed to get dtc if not documented here.

It's not required; 23df708aa36d9b748b54d1c362e7e2e67316593c

Author: > I'm sure this is documented somewhere 

https://docs.zephyrproject.org/latest/guides/dts/intro.html#scripts-and-tools

Reviewer 4: Ho, right! 
I was confused about dtc status, I knew it was still in use (at least on my side).
That's clear now. Txs @mbolivar-nordic

Export Zephyr CMake package
***************************

The :ref:`Zephyr CMake package <cmake_pkg>` allows CMake to load boilerplate

> code required for building Zephyr applications.

Reviewer: Nothing wrong here, but even if you don't export, then CMake can still load the boilerplate code when using `ZEPHYR_BASE`.
So the main benefit here is 
```suggestion
The :ref:`Zephyr CMake package <cmake_pkg>` allows CMake to automatically load boilerplate
code required for building Zephyr applications.
```

         .. code-block:: bash

>             cd ~/zephyrproject

Reviewer: Personally, I prefer to have my tools / programs outside any repository.

Especially as tools might be up-/downgrade independently of the repo.

Or I could have multiple up-/downstream workspaces, but just a single toolchain.

So I would suggest to still advice users with a different path.

Which of course also means the comment goes for remaining part of this commit.

Author: I am doing this on purpose for the getting started guide precisely to make the Zephyr environment more self-contained, so we appear to have different preferences in this regard. The question is really what will be easier for more people, I suppose.

Reviewer: could you take a look at:
https://github.com/zephyrproject-rtos/sdk-ng/pull/199
https://github.com/zephyrproject-rtos/zephyr/pull/24143

those will remove the need for the environment variables, and in such way make it easier to get started.
And not polluting the Zephyr workspace with a toolchain folder.

Author: > could you take a look at:
> [zephyrproject-rtos/sdk-ng#199](https://github.com/zephyrproject-rtos/sdk-ng/pull/199)
> #24143
> 

OK, I like that PR very much. Let's wait until it gets reviewed by @nashif and @galak to decide how to proceed here.

Reviewer 2: I suggest using either `/opt/` or ~/bin/` instead

Reviewer: > I suggest using either `/opt/` or ~/bin/` instead

I think `~/bin` is a bad option, as the Zephyr SDK is not a single executable file, but several folders with executables, libraries, etc.

Currently, the suggested default paths to search for the SDK are:
```
$HOME/zephyr-sdk*
$HOME/.local/zephyr-sdk*
$HOME/.local/opt/zephyr-sdk*
/opt/zephyr-sdk*
/usr/zephyr-sdk*
/usr/local/zephyr-sdk*
```
so I suggest we pick one among those in the example given in gsg.

Feel free to comment on default paths for SDK in this PR:
https://github.com/zephyrproject-rtos/sdk-ng/pull/199

Reviewer 2: I meant `~/bin/zephyr-sdk` @tejlmand
